cfme/tests/cloud/test_cloud_init_provisioning.py

```diff
@@ -55,20 +61,50 @@ def test_provision_cloud_init(request, setup_provider, provider, provisioning,
     instance = Instance.factory(vm_name, provider, image)
 
     request.addfinalizer(instance.cleanup_on_provider)
-    # TODO: extend inst_args for other providers except EC2 if needed
     inst_args = {
         'request': {'email': 'image_provisioner@example.com',
                     'first_name': 'Image',
                     'last_name': 'Provisioner',
                     'notes': note},
-        'catalog': {'vm_name': vm_name},
-        'properties': {'instance_type': provisioning['instance_type'],
-                       'guest_keypair': provisioning['guest_keypair']},
-        'environment': {'availability_zone': provisioning['availability_zone'],
-                        'cloud_network': provisioning['cloud_network'],
-                        'security_groups': [provisioning['security_group']]},
-        'customize': {'custom_template': {'name': provisioning['ci-template']}}
+        'catalog': {
+            'vm_name': vm_name},
+        'customize': {
+            'custom_template': {'name': provisioning['ci-template']}},
+        'environment': {
+            'availability_zone': provisioning.get('availability_zone', None),
+            'security_groups': [provisioning.get('security_group', None)],
+            'cloud_network': provisioning.get('cloud_network', None),
+            'cloud_subnet': provisioning.get('cloud_subnet', None),
+            'resource_groups': provisioning.get('resource_group', None)
+        },
+        'properties': {
+            'instance_type': partial_match(provisioning.get('instance_type', None)),
+            'guest_keypair': provisioning.get('guest_keypair', None)}
     }
+    # GCE specific
+    if provider.one_of(GCEProvider):
+        recursive_update(inst_args, {
+            'properties': {
+                'boot_disk_size': provisioning['boot_disk_size'],
+                'is_preemptible': True}
+        })
+
+    # Azure specific
+    if provider.one_of(AzureProvider):
+        # Azure uses different provisioning keys for some reason
+        try:
+            template = provider.data.templates.small_template
+            vm_user = credentials[template.creds].username
+            vm_password = credentials[template.creds].password
+        except AttributeError:
+            pytest.skip('Could not find small_template or credentials for {}'.format(provider.name))
```

Comment:
@izapolsk I think this is the correct behavior for our tests - we want to run some test that requires a small_template be defined for the provider. If you have some incomplete yaml, the test doesn't fail with some AttributeError/KeyError, etc, it skips. This counts against your overall pass %, but doesn't appear as though some functionality is broken or some test implemented incorrectly.

Taking this to the extreme - if you tried to collect and run tests with only the absolutely bare minimum provider definition in yaml, you would end up with many (thousands) of skipped tests, with each recording what field was expected and missing.

It makes it more explicit that there weren't any _failures_ in this case, but missing data preventing the test (or its setup) from actually running.

Reply:
@mshriver, I believe there shouldn't be incomplete yamls. 
1. If test doesn't support some provider, there should be obvious uncollect/skip statement.
2. If test data is incomplete, test should fail not skip. Because we won't see this test data bug otherwise.